File: apidemo/sensordemo/motionsensordemo/src/main/java/com/gdkdemo/sensor/motion/service/MotionSensorDemoLocalService.java
package com.gdkdemo.sensor.motion.service;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Binder;
import android.os.IBinder;
import android.widget.RemoteViews;

import com.gdkdemo.sensor.motion.MotionSensorDemoActivity;
import com.gdkdemo.sensor.motion.R;
import com.gdkdemo.sensor.motion.common.SensorValueStruct;
import com.google.android.glass.timeline.LiveCard;
import com.google.android.glass.timeline.TimelineManager;


// ...
public class MotionSensorDemoLocalService extends Service implements SensorEventListener
{
    // For live card
    private LiveCard liveCard = null;

    // Sensor manager
    private SensorManager mSensorManager = null;

    // Motion sensors
    private Sensor mSensorAccelerometer = null;
    private Sensor mSensorGravity = null;
    private Sensor mSensorLinearAcceleration = null;
    private Sensor mSensorGyroscope = null;
    private Sensor mSensorRotationVector = null;


    // Last known motionsensor values.
    private SensorValueStruct lastSensorValuesAccelerometer = null;
    private SensorValueStruct lastSensorValuesGravity = null;
    private SensorValueStruct lastSensorValuesLinearAcceleration = null;
    private SensorValueStruct lastSensorValuesGyroscope = null;
    private SensorValueStruct lastSensorValuesRotationVector = null;

    // TBD:
    // Need a timer, etc. to refresh the UI (LiveCard) ever x seconds, etc..
    // For now, we just use the "last timestamp".
    private long lastRefreshedTime = 0L;


    // No need for IPC...
    public class LocalBinder extends Binder {
        public MotionSensorDemoLocalService getService() {
            return MotionSensorDemoLocalService.this;
        }
    }
    private final IBinder mBinder = new LocalBinder();


    @Override
    public void onCreate()
    {
        super.onCreate();
        initializeSensorManager();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId)
    {
        Log.i("Received start id " + startId + ": " + intent);
        onServiceStart();
        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent)
    {
        // ????
        onServiceStart();
        return mBinder;
    }

    @Override
    public void onDestroy()
    {
        // ???
        onServiceStop();
        super.onDestroy();
    }


    @Override
    public void onSensorChanged(SensorEvent event)
    {
        Log.d("onSensorChanged() called.");

        processMotionSensorData(event);

    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy)
    {
        Log.d("onAccuracyChanged() called.");

    }



    // Service state handlers.
    // ....

    private boolean onServiceStart()
    {
        Log.d("onServiceStart() called.");

        mSensorManager.registerListener(this, mSensorAccelerometer, SensorManager.SENSOR_DELAY_NORMAL);
        mSensorManager.registerListener(this, mSensorGravity, SensorManager.SENSOR_DELAY_NORMAL);
        mSensorManager.registerListener(this, mSensorLinearAcceleration, SensorManager.SENSOR_DELAY_NORMAL);
        mSensorManager.registerListener(this, mSensorGyroscope, SensorManager.SENSOR_DELAY_NORMAL);
        mSensorManager.registerListener(this, mSensorRotationVector, SensorManager.SENSOR_DELAY_NORMAL);

        // Publish live card...
        publishCard(this);

        return true;
    }

    private boolean onServicePause()
    {
        Log.d("onServicePause() called.");
        return true;
    }
    private boolean onServiceResume()
    {
        Log.d("onServiceResume() called.");
        return true;
    }

    private boolean onServiceStop()
    {
        Log.d("onServiceStop() called.");

        mSensorManager.unregisterListener(this);

        // TBD:
        // Unpublish livecard here
        // .....
        unpublishCard(this);
        // ...

        return true;
    }




    // For live cards...

    private void publishCard(Context context)
    {
        publishCard(context, false);
    }
    private void publishCard(Context context, boolean update)
    {
        if(Log.D) Log.d("publishCard() called: update = " + update);
        if (liveCard == null || update == true) {

//            // TBD:
//            // We get multiple liveCards if we just call setViews() and publish()...
//            // As a workaround, for now, we always unpublish the previous card first.
//            if (liveCard != null) {
//                liveCard.unpublish();
//            }
//            // ....

            final String cardId = "motionsensordemo_card";
            // Note: getLiveCard() always publishes a new live card....
            if(liveCard == null) {
            // if(liveCard == null || ! liveCard.isPublished()) {
                TimelineManager tm = TimelineManager.from(context);
                liveCard = tm.createLiveCard(cardId);
//                 liveCard.setNonSilent(true);       // for testing.
            }
            RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.livecard_motionsensordemo);
            String content = "";
            if(lastSensorValuesAccelerometer != null) {
                content += "Accelerometer:\n" + lastSensorValuesAccelerometer.toString() + "\n";
            }
            if(lastSensorValuesGravity != null) {
                content += "Gravity:\n" + lastSensorValuesGravity.toString() + "\n";
            }
            if(lastSensorValuesLinearAcceleration != null) {
                content += "Linear Acceleration:\n" + lastSensorValuesLinearAcceleration.toString() + "\n";
            }
            if(lastSensorValuesGyroscope != null) {
                content += "Gyroscope:\n" + lastSensorValuesGyroscope.toString() + "\n";
            }
            if(lastSensorValuesRotationVector != null) {
                content += "Rotation Vector:\n" + lastSensorValuesRotationVector.toString() + "\n";
            }

            remoteViews.setCharSequence(R.id.livecard_content, "setText", content);
            liveCard.setViews(remoteViews);
            Intent intent = new Intent(context, MotionSensorDemoActivity.class);
            liveCard.setAction(PendingIntent.getActivity(context, 0, intent, 0));
            // ???
            // Without this if(),
            // I get an exception:
            // "java.lang.IllegalStateException: State CREATED expected, currently in PUBLISHED"
            // Why???
            if(! liveCard.isPublished()) {
                liveCard.publish(LiveCard.PublishMode.REVEAL);
            } else {
                // ????
                // According to the GDK doc,
                // it appears we should call publish() every time the content changes...
                // But, it seems to work without re-publishing...
                if(Log.D) {
                    long now = System.currentTimeMillis();
                    Log.d("liveCard not published at " + now);
                }
            }
        } else {
            // Card is already published.
            return;
        }
    }

    private void unpublishCard(Context context)
    {
        Log.d("unpublishCard() called.");
        if (liveCard != null) {
            liveCard.unpublish();
            liveCard = null;
        }
    }


    // MotionSensor methods
    private void initializeSensorManager()
    {
        mSensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);
        mSensorAccelerometer = mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        mSensorGravity = mSensorManager.getDefaultSensor(Sensor.TYPE_GRAVITY);
        mSensorLinearAcceleration = mSensorManager.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION);
        mSensorGyroscope = mSensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE);
        mSensorRotationVector = mSensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
    }

    private void processMotionSensorData(SensorEvent event)
    {
        long now = System.currentTimeMillis();

        Sensor sensor = event.sensor;
        int type = sensor.getType();
        long timestamp = event.timestamp;
        float[] values = event.values;
        int accuracy = event.accuracy;
        SensorValueStruct data = new SensorValueStruct(type, timestamp, values, accuracy);

        switch(type) {
            case Sensor.TYPE_ACCELEROMETER:
                lastSensorValuesAccelerometer = data;
                break;
            case Sensor.TYPE_GRAVITY:
                lastSensorValuesGravity = data;
                break;
            case Sensor.TYPE_LINEAR_ACCELERATION:
                lastSensorValuesLinearAcceleration = data;
                break;
            case Sensor.TYPE_GYROSCOPE:
                lastSensorValuesGyroscope = data;
                break;
            case Sensor.TYPE_ROTATION_VECTOR:
                lastSensorValuesRotationVector = data;
                break;
            default:
                Log.w("Unknown type: " + type);
        }

        // TBD:
        // Update the DB, etc..
        // ...

        // Update the UI.
        // temporary
        final long delta = 5000L;   // every 5 seconds
        if(lastRefreshedTime <= now - delta) {
            // if(liveCard != null && liveCard.isPublished()) {
                // Update...
                publishCard(this, true);
            // }
            lastRefreshedTime = now;
        }
    }


}

File: xamarindemo/sensordemo/MotionSensorDemo/MotionSensorDemo/service/MotionSensorDemoLocalService.cs
using System;
using Android.App;
using Android.Content;
using Android.Glass.App;
using Android.Glass.Touchpad;
using Android.Hardware;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Android.Glass.Timeline;
using System.Threading;
using Android.Util;
using Android.Database;
using System.Collections.Generic;
using System.Linq;

namespace MotionSensorDemo
{
	public class MotionSensorDemoLocalService : Service, ISensorEventListener
	{
		public MotionSensorDemoLocalService () :base()
		{
		}

		private string _tag = "EnvironmentalSensorDemoLocalService";

		// For live card
		private LiveCard liveCard = null;

		// Sensor manager
		private SensorManager mSensorManager = null;

		// Motion sensors
		private Sensor mSensorAccelerometer = null;
		private Sensor mSensorGravity = null;
		private Sensor mSensorLinearAcceleration = null;
		private Sensor mSensorGyroscope = null;
		private Sensor mSensorRotationVector = null;


		// Last known motionsensor values.
		private SensorValueStruct lastSensorValuesAccelerometer = null;
		private SensorValueStruct lastSensorValuesGravity = null;
		private SensorValueStruct lastSensorValuesLinearAcceleration = null;
		private SensorValueStruct lastSensorValuesGyroscope = null;
		private SensorValueStruct lastSensorValuesRotationVector = null;

		// TBD:
		// Need a timer, etc. to refresh the UI (LiveCard) ever x seconds, etc..
		// For now, we just use the "last timestamp".
		private long lastRefreshedTime = 0L;


		// No need for IPC...
		public class LocalBinder : Binder {
			MotionSensorDemoLocalService _service;

			public LocalBinder (MotionSensorDemoLocalService service)
			{
				_service = service;
			}

			public MotionSensorDemoLocalService GetService() {
				return _service;
			}
		}
		private IBinder mBinder = new LocalBinder(new MotionSensorDemoLocalService());


		public override void OnCreate()
		{
			base.OnCreate();
			InitializeSensorManager();
		}

		public StartCommandResult OnStartCommand(Intent intent, StartCommandFlags flags, int startId)
		{
			Log.Info(_tag, "Received start id " + startId + ": " + intent);
			OnServiceStart();
			return StartCommandResult.Sticky;
		}

		public override IBinder OnBind(Intent intent)
		{
			// ????
			OnServiceStart();
			return mBinder;
		}

		public override void OnDestroy()
		{
			OnServiceStop();
			base.OnDestroy();
		}


		public void OnSensorChanged(SensorEvent e)
		{
			Log.Debug(_tag, "onSensorChanged() called.");

			ProcessMotionSensorData(e);

		}

		public void OnAccuracyChanged(Sensor sensor, SensorStatus accuracy)
		{
			Log.Debug(_tag, "onAccuracyChanged() called.");
		}

		// Service state handlers.
		// ....

		private bool OnServiceStart()
		{
			Log.Debug(_tag, "onServiceStart() called.");

			mSensorManager.RegisterListener(this, mSensorAccelerometer, SensorDelay.Normal);
			mSensorManager.RegisterListener(this, mSensorGravity, SensorDelay.Normal);
			mSensorManager.RegisterListener(this, mSensorLinearAcceleration, SensorDelay.Normal);
			mSensorManager.RegisterListener(this, mSensorGyroscope, SensorDelay.Normal);
			mSensorManager.RegisterListener(this, mSensorRotationVector, SensorDelay.Normal);

			// Publish live card...
			PublishCard(this);

			return true;
		}

		private bool OnServicePause()
		{
			Log.Debug(_tag, "onServicePause() called.");
			return true;
		}
		private bool OnServiceResume()
		{
			Log.Debug(_tag, "onServiceResume() called.");
			return true;
		}

		private bool OnServiceStop()
		{
			Log.Debug(_tag, "onServiceStop() called.");

			mSensorManager.UnregisterListener(this);

			// TBD:
			// Unpublish livecard here
			// .....
			UnpublishCard(this);
			// ...

			return true;
		}




		// For live cards...

		private void PublishCard(Context context)
		{
			publishCard(context, false);
		}

		private void publishCard(Context context, bool update)
		{
			Log.Debug(_tag, "publishCard() called: update = " + update);
			if (liveCard == null || update == true) {

//            // TBD:
//            // We get multiple liveCards if we just call setViews() and publish()...
//            // As a workaround, for now, we always unpublish the previous card first.
//            if (liveCard != null) {
//                liveCard.unpublish();
//            }
//            // ....

				string cardId = "motionsensordemo_card";
				// Note: getLiveCard() always publishes a new live card....
				if(liveCard == null) {
					// if(liveCard == null || ! liveCard.isPublished()) {
					TimelineManager tm = TimelineManager.From(context);
					liveCard = tm.CreateLiveCard(cardId);
//                 liveCard.setNonSilent(true);       // for testing.
				}
				RemoteViews remoteViews = new RemoteViews(context.PackageName, Resource.Layout.LiveCard_MotionSensorDemo);
				String content = "";
				if(lastSensorValuesAccelerometer != null) {
					content += "Accelerometer:\n" + lastSensorValuesAccelerometer.ToString() + "\n";
				}
				if(lastSensorValuesGravity != null) {
					content += "Gravity:\n" + lastSensorValuesGravity.ToString() + "\n";
				}
				if(lastSensorValuesLinearAcceleration != null) {
					content += "Linear Acceleration:\n" + lastSensorValuesLinearAcceleration.ToString() + "\n";
				}
				if(lastSensorValuesGyroscope != null) {
					content += "Gyroscope:\n" + lastSensorValuesGyroscope.ToString() + "\n";
				}
				if(lastSensorValuesRotationVector != null) {
					content += "Rotation Vector:\n" + lastSensorValuesRotationVector.ToString() + "\n";
				}

				remoteViews.SetCharSequence(Resource.Id.livecard_content, "setText", content);
				liveCard.SetViews(remoteViews);
				Intent intent = new Intent(context, typeof(MotionSensorDemoActivity));
				liveCard.SetAction(PendingIntent.GetActivity(context, 0, intent, 0));
				// ???
				// Without this if(),
				// I get an exception:
				// "java.lang.IllegalStateException: State CREATED expected, currently in PUBLISHED"
				// Why???
				if(! liveCard.IsPublished) {
					liveCard.Publish(LiveCard.PublishMode.Reveal);
				} else {
					// ????
					// According to the GDK doc,
					// it appears we should call publish() every time the content changes...
					// But, it seems to work without re-publishing...
					long now = DateTime.UtcNow.Ticks;
					Log.Debug(_tag, "liveCard not published at " + now);
				}
			} else {
				// Card is already published.
				return;
			}
		}

		private void UnpublishCard(Context context)
		{
			Log.Debug(_tag, "unpublishCard() called.");
			if (liveCard != null) {
				liveCard.Unpublish();
				liveCard = null;
			}
		}

		// MotionSensor methods
		private void InitializeSensorManager()
		{
			mSensorManager = (SensorManager) GetSystemService(Context.SensorService);
			mSensorAccelerometer = mSensorManager.GetDefaultSensor(SensorType.Accelerometer);
			mSensorGravity = mSensorManager.GetDefaultSensor(SensorType.Gravity);
			mSensorLinearAcceleration = mSensorManager.GetDefaultSensor(SensorType.LinearAcceleration);
			mSensorGyroscope = mSensorManager.GetDefaultSensor(SensorType.Gyroscope);
			mSensorRotationVector = mSensorManager.GetDefaultSensor(SensorType.RotationVector);
		}

		private void ProcessMotionSensorData(SensorEvent e)
		{
			long now = DateTime.UtcNow.Ticks;

			SensorType sensor = e.Sensor.Type;
			//Type type = sensor.GetType ();
			long timestamp = e.Timestamp;
			IList<float> values = e.Values;
			SensorStatus accuracy = e.Accuracy;
			SensorValueStruct data = new SensorValueStruct (sensor, timestamp, values, accuracy);

			switch (sensor) {
			case SensorType.Accelerometer:
				lastSensorValuesAccelerometer = data;
				break;
			case SensorType.Gravity:
				lastSensorValuesGravity = data;
				break;
			case SensorType.LinearAcceleration:
				lastSensorValuesLinearAcceleration = data;
				break;
			case SensorType.Gyroscope:
				lastSensorValuesGyroscope = data;
				break;
			case SensorType.RotationVector:
				lastSensorValuesRotationVector = data;
				break;
			default:
				Log.Warn (_tag, "Unknown type: " + sensor);
				break;
			}

			// TBD:
			// Update the DB, etc..
			// ...

			// Update the UI.
			// temporary
			long delta = 5000L;   // every 5 seconds
			if (lastRefreshedTime <= now - delta) {
				// if(liveCard != null && liveCard.isPublished()) {
				// Update...
				publishCard (this, true);
				// }
				lastRefreshedTime = now;
			}
		}
	}
}

